File: FlyingDanger/FlyingDangerC++/Enemy.cpp
#include "Enemy.hpp"
#include "GameManager.hpp"
#include "Castle.hpp"

int currentMeteorHealth = 100;
int maxMeteorHealth = 100;
int defaultSpeed = 2;
int defaultDamage = 10;


Enemy::Enemy() : health(currentMeteorHealth), max_health(maxMeteorHealth), speed(defaultSpeed), position(""), sprite(""), rect(""), game_manager(nullptr) {
    cout << "Объект класса Enemy создан" << endl;
}

Enemy::Enemy(int hp, int spd) : health(hp), max_health(hp), speed(spd), position(""), sprite(""), rect(""),
game_manager(nullptr) {
    cout << "Объект класса Enemy создан с параметрами" << endl;
}

Enemy::Enemy(int hp, int spd, GameManager* gm) : health(hp), max_health(hp), speed(spd), position(""),
sprite(""), rect(""), game_manager(gm) {
    cout << "Объект класса Enemy создан с ссылкой на GameManager" << endl;
}

void Enemy::AttackCastle(Castle* castle) {
    if (castle && IsAlive()) {
        cout << "Враг атакует замок!" << endl;
        castle->TakeDamage(defaultDamage);
    }
}

File: src/Enemy.java
public abstract class Enemy {
    protected int health;
    protected int maxHealth;
    protected int speed;
    protected String position;
    protected String sprite;
    protected String rect;
    protected GameManager gameManager;

    // Константы
    public static final int DEFAULT_HEALTH = 50;
    public static final int DEFAULT_MAX_HEALTH = 50;
    public static final int DEFAULT_SPEED = 2;
    public static final int CASTLE_ATTACK_DAMAGE = 10;

    public Enemy() {
        this.health = DEFAULT_HEALTH;
        this.maxHealth = DEFAULT_MAX_HEALTH;
        this.speed = DEFAULT_SPEED;
        this.position = "";
        this.sprite = "";
        this.rect = "";
        this.gameManager = null;
        System.out.println("Объект класса Enemy создан");
    }

    public Enemy(int hp, int spd) {
        this();
        this.health = hp;
        this.maxHealth = hp;
        this.speed = spd;
        System.out.println("Объект класса Enemy создан с параметрами");
    }

    public Enemy(int hp, int spd, GameManager gm) {
        this(hp, spd);
        this.gameManager = gm;
        System.out.println("Объект класса Enemy создан с ссылкой на GameManager");
    }

    public abstract void move();
    public abstract void takeDamage(int amount);
    public abstract boolean checkCollision(String target);
    public abstract void update();
    public abstract void draw();

    public void attackCastle(Castle castle) {
        if (castle != null && isAlive()) {
            System.out.println("Враг атакует замок!");
            castle.takeDamage(CASTLE_ATTACK_DAMAGE);
        }
    }

    public boolean isAlive() {
        return health > 0;
    }

    public void checkBoundaries() {
        System.out.println("Базовый метод checkBoundaries() класса Enemy");
    }

    public int getMaxHealth() { return maxHealth; }
}